java/109-triangle.md

```diff
@@ -8,6 +8,148 @@
+        //return 
+        if(x == triangle.length){
+            min = Math.min(min, sum);
+            return;
+        }
+        
+        
+        traverse(triangle, x + 1, y, sum + triangle[x][y]);
+        traverse(triangle, x + 1, y + 1, sum + triangle[x][y]);
+    }
+}
+
+```
+**DFS + memorization**
+```
+public class Solution {
+    /**
+     * @param triangle: a list of lists of integers.
+     * @return: An integer, minimum path sum.
+     */
+    public int minimumTotal(int[][] triangle) {
+        // write your code here
+        int n = triangle.length;
+        int [][] hash = new int[n][n];
+        //Initializate
+        for(int i = 0; i < n; i++){
+            for(int j = 0; j < n; j++){
+                hash[i][j] = Integer.MAX_VALUE;
+            }
+        }
+        return divideConquer(triangle, hash, 0, 0);
+    }
+    //minimum path value from x,y to bottom
+    private int divideConquer(int[][] triangle,int[][]hash, int x,int y){
+        //boundary case
+        int n = triangle.length;
+        if(x  == n - 1){
+            return triangle[x][y];
+        }
+        //or
+        // if(x == n){
+        //     return 0;
+        // }
+        if(hash[x][y] != Integer.MAX_VALUE){
+            return hash[x][y];
+        }
+        
+        hash[x][y] = triangle[x][y] + Math.min(divideConquer(triangle, hash, x + 1, y),divideConquer(triangle, hash, x + 1,y + 1));
+        return hash[x][y];
+    }
+}
+```
+**Bottom-up**
+```
+public class Solution {
+    /**
+     * @param triangle: a list of lists of integers.
+     * @return: An integer, minimum path sum.
+     */
+    public int minimumTotal(int[][] triangle) {
+        // write your code here
+        if(triangle == null || triangle.length == 0){
+            return -1;
+        }
+        if(triangle[0] == null ||triangle[0].length == 0 ){
+            return -1;
+        }
+        
+        int n = triangle.length;
+        int [][] f = new int [n][n];//minimum path from x,y to bottom
+        // Initialize
+        for (int i = 0; i < n; i++){
+            f[n - 1][i] = triangle[n - 1][i];
+        }
+        
+        for(int i = n - 2; i >= 0; i--){
+            for(int j = 0; j <= i; j++){
+                f[i][j] = Math.min(f[i + 1][j], f[i + 1][j + 1]) + triangle[i][j];
+            }
+        }
+        return f[0][0];
+    }
+}
+
+```
+**Top down**
+*caution:the initialization of i and j !  not 0*
+```
+public class Solution {
+    /**
+     * @param triangle: a list of lists of integers.
+     * @return: An integer, minimum path sum.
+     */
+    public int minimumTotal(int[][] triangle) {
+        // write your code here
+        if(triangle == null || triangle.length == 0){
+            return -1;
+        }
+        if(triangle[0] == null || triangle[0].length == 0){
+            return -1;
+        }
+        //State :minimum path from 0,0 to x,y
+        int n = triangle.length;
+        int [][] f = new int[n][n];
```

Comment:
Here we may can use rotate array to achieve O(n) space complexity.
e.g. `int [][] f = new int[2][n];`

and in loop we could have some thing like:
```java
f[i % 2][j] = f[ (i - 1) % 2][j] ....
```

Just a little improvement in case of interviewer asking for O(n) space optimization:)